### ISSUE TITLE
DQL does not work when setlocale(LC_ALL, 'tr_TR'); was called before

### ISSUE THREAD

**Issue by Arany** [doctrine/doctrine2#5564]
When you set the locale with `setlocale` for specific languages before using DQL, the Lexer splits the tokens wrong:

(inside a ProductGroupMemberRepository EntityRepository class)
```php
    setlocale(LC_ALL, 'tr_TR');

    $qb   = $this->createQueryBuilder('m');
        $expr = $qb->expr();

        $qb->select('p')
            ->innerJoin('Entity\Product', 'p', 'WITH', 'm.productId = p.id')
            ->where($expr->eq("m.groupId", ":groupId"))
            ->setParameter(":groupId", $groupId);
    return $qb->getQuery()->getResult();
```

It breaks with the error message: `Error: Expected end of string, got 'I'`
After some debugging, I found that preg_split apparently works differently on tr_TR. it splits on different  characters: 

```
array(36) {
  [0]=>
  array(2) {
    [0]=>
    string(6) "SELECT"
    [1]=>
    int(0)
  }
  [1]=>
  array(2) {
    [0]=>
    string(1) "p"
    [1]=>
    int(7)
  }
  [2]=>
  array(2) {
    [0]=>
    string(4) "FROM"
    [1]=>
    int(9)
  }
  [3]=>
  array(2) {
    [0]=>
    string(40) "Entity\ProductGroupMember"
    [1]=>
    int(14)
  }
  [4]=>
  array(2) {
    [0]=>
    string(1) "m"
    [1]=>
    int(55)
  }
  [5]=>
  array(2) {
    [0]=>
    string(1) "I"
    [1]=>
    int(57)
  }
  [6]=>
  array(2) {
    [0]=>
    string(4) "NNER"
    [1]=>
    int(58)
  }
  [7]=>
  array(2) {
    [0]=>
    string(2) "JO"
    [1]=>
    int(63)
  }
  [8]=>
  array(2) {
    [0]=>
    string(1) "I"
    [1]=>
    int(65)
  }
  [9]=>
  array(2) {
    [0]=>
    string(1) "N"
    [1]=>
    int(66)
  },
...
```

If I use `setlocale` with `LC_NUMERIC` or comment the line out, or use other languages, everything works fine. Even though the regex and DQL are exactly the same thing.

**Comment by Ocramius** [comment]
Any pointers as to where the execution path in doctrine/lexer changes (when the locale is changed)?

**Comment by zeroedin-bill** [comment]
I just wrote a test (#5565) for the lexer with a bunch of locales and it seems to be fine on Windows with PHP 5.6.13. 

@Arany What OS and PHP version are you using?

**Comment by Arany** [comment]
It is the same exexution path. The Problem appelarently lies in preg_split that takes the same regex and input. The only difference being the setlocale..

I tested it on different machines and now it looks to me like a php-bug. On the machine with 5.5.18 it breaks. On 5.5.30 everything is fine. Only that i do not have Access to the php Installation on the Server so i cannot Upgrade it..

**Comment by Arany** [comment]
The Server is using CentOS.

**Comment by zeroedin-bill** [comment]
@Arany I really don't think there's anything we can do in Doctrine core to mitigate this php bug. If the preg_* functions are behaving badly in the lexer, they're probably behaving badly elsewhere, which could lead to serious issues elsewhere in your app. There were some nasty locale thread safety bugs fixed in 5.5.21.

Your best bet is to get the administrator of the machine to update the PHP installation.

**Comment by halk** [comment]
I just had the same error with PHP 5.6.18. Must be related to the host, as all is fine on my mac. The server runs Ubuntu.
